src/GeositeFramework/plugins/layer_selector_v2/State.js

```diff
@@ -90,166 +19,47 @@ define([
-                        return layer.getDisplayName().toLowerCase().indexOf(filterText) !== -1;
+            toggleLayer: function(layer) {
+                var layerId = layer.id();
+                if (layer.isFolder()) {
+                    if (this.isExpanded(layerId)) {
+                        return this.collapseLayer(layerId);
                     } else {
-                        layer.children = _.filter(layer.children, filterLayer);
-                        return layer.getChildren().length > 0;
+                        return this.expandLayer(layerId);
                     }
-                });
-            },
-
-            filterByRegion: function(layers, currentRegion) {
-                return _.filter(layers, function filterLayer(layer) {
-                    if (!layer.isAvailableInRegion(currentRegion)) {
-                        return false;
-                    } else if (layer.isFolder()) {
-                        layer.children = _.filter(layer.children, filterLayer);
-                        return layer.getChildren().length > 0;
+                } else {
+                    if (this.isSelected(layerId)) {
+                        return this.deselectLayer(layerId);
+                    } else {
+                        return this.selectLayer(layerId);
                     }
-                    return true;
-                });
-            },
-
-            // Ensure that `layers` always reflects the union of configuration
-            // data with data fetched from map services.
-            rebuildLayers: function() {
-                this.layers = this.coalesceLayers();
-
-                // `coalesceLayers` needs to be executed again because the filter
-                // logic mutates the data and we don't want to alter `layers`.
-                var filteredLayers = this.coalesceLayers();
-                filteredLayers = this.filterByRegion(filteredLayers, this.currentRegion);
-                filteredLayers = this.filterByName(filteredLayers, this.getFilterText());
-                this.filteredLayers = filteredLayers;
-
-                this.emit(LAYERS_CHANGED);
-            },
-
-            getLayers: function() {
-                return this.filteredLayers;
-            },
-
-            findLayer: function(layerId) {
-                return util.find(this.layers, function(layer) {
-                    return layer.findLayer(layerId);
-                });
+                }
             },
 
             getSelectedLayers: function() {
-                return _.map(this.savedState.selectedLayers, this.findLayer, this);
-            },
-
-            clearAll: function() {
-                this.filterTree('');
-                this.setSelectedLayers([]);
-                this.setExpandedLayers([]);
-                this.emit(EVERYTHING_CHANGED);
+                return this.savedState.selectedLayers;
             },
 
             isSelected: function(layerId) {
```

Comment:
Yeah. Unfortunately, it's very easy to (potentially) get out of sync by forgetting to update references. I'll try to think about design patterns we can use to prevent this from happening.